### PR TITLE
Types support

### LINKED ISSUE #5: Add support for Date/Time/Timestamp types

### LINKED ISSUE #6: Add implicit type conversions for JdbcResultSet when possible
For example, if we have a column of type `int`, and then we call `getLong`, then currently we will get a ClassCastException but in fact, it should safely convert it to `long`

### LINKED ISSUE #18: Add implicit type conversions for PreparedStatement parameters when possible

### PR DESCRIPTION
**Pull request by SeriyBg**
Fix #6
Fix #5 
Fix #18

### CODE REVIEW

File: src/main/java/com/hazelcast/jdbc/JdbcResultSet.java [hunk 1]
```diff
@@ -110,42 +112,42 @@ public boolean wasNull() throws SQLException {
...
+        return convertTo(get(columnIndex), Float.class);
     }
 
     @Override
     public double getDouble(int columnIndex) throws SQLException {
-        return get(columnIndex);
+        return convertTo(get(columnIndex), Double.class);
     }
 
     @Override
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:153]: `BigDecimal` is one of our built-it types, so we should support it as well.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        CLASS_TO_QUERY_TYPE.put(String.class, QueryDataTypeFamily.VARCHAR);
+        CLASS_TO_QUERY_TYPE.put(Boolean.class, QueryDataTypeFamily.BOOLEAN);
+        CLASS_TO_QUERY_TYPE.put(BigDecimal.class, QueryDataTypeFamily.DECIMAL);
+        CLASS_TO_QUERY_TYPE.put(Timestamp.class, QueryDataTypeFamily.TIMESTAMP_WITH_TIME_ZONE);
+        CLASS_TO_QUERY_TYPE.put(Time.class, QueryDataTypeFamily.TIME);
+        CLASS_TO_QUERY_TYPE.put(Date.class, QueryDataTypeFamily.DATE);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T convertTo(Object object, Class<T> clazz) throws SQLException {
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/TypeConverter.java:78]: Getting the value from the driver is a performance-sensitive operation. The current implementation looks too heavy to me, especially for simple types:

1. For all methods except `getObject`, `getDate`, `getTime` and `getTimestamp`, we do not need to do any lookups. It should be sufficient to do the following. It requires only one hash map lookup as opposed to three in the current impl:
```
static <T> T convertTo(Object object, QueryDataType type) throws SQLException {
    try {
         return type.convert(object);
    } catch (Exception e) {
        // handle exception
    }
        }
}
```
2. For date, time, timestamp: first invoke the method from p.1, then apply the subsequent conversion. One lookup instead of three.
3. For object: 
- resolve the type of the passed class using `QueryDataTypeUtils.resolveTypeForClass`
- invoke logic from p.1
- if requested class is one of `java.sql.Date`, `java.sql.Time`, `java.sql.Timestamp` - invoke logic from p.2
- check if the produced object is assignable to the requested class. If yes - return, otherwise - throw a proper exception (currently we will throw an implicit `ClassCastException`)

This way we have simple implementation for simple cases, and gradually make it heavier for more complex cases. We also eliminate two of three hash map lookups.

File: src/main/java/com/hazelcast/jdbc/JdbcResultSet.java [hunk 3]
```diff
@@ -43,6 +43,8 @@
 import java.util.Iterator;
 import java.util.Map;
 
+import static com.hazelcast.jdbc.TypeConverter.convertTo;
+
 public class JdbcResultSet implements ResultSet {
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:48]: `getObject(int columnIndex, Class<T> type)` and `getObject(String columnLabel, Class<T> type)` methods ignore the passed type.

Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:48]: According to JDBC spec, a somewhat similar approach to conversions is required for parameters (`PreparedStatement.setObject(...)`). Do we have it on our radars?

File: src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Date;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class DriverTypeConversionTest {
```
Comment (devozerov) [inline src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java:49]: To me, the test coverage looks insufficient. Are we sure that we test every possible pair of `from/to` types? This is tedious, but getting the results from the `ResultSet` is a foundational part of the driver, and we can't allow bugs here. 
I would propose to test every type pair separately with a precise contract: 
1. If it should succeed, then what is the expected result?
2. If it should fail, then check that only `SQLException` is thrown, and check for the specific error message (since error messages are part of the public API).

We did exactly the same for the server part. See `com.hazelcast.sql.impl.type.converter.ConvertersTest` and `com.hazelcast.sql.impl.type.converter.CastFunctionIntegrationTest` as examples of this exhaustive testing approach. Also, there is a set of convenient classes to return different types of values from the query: `com.hazelcast.sql.support.expressions.ExpressionTypes`, `com.hazelcast.sql.support.expressions.ExpressionValue`.

File: src/main/java/com/hazelcast/jdbc/JdbcResultSet.java (hunk 3 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:48]: Created an issue #18

File: src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java (hunk 4 above)
Comment (SeriyBg) [reply src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java:54]: Have added move test cases to cover from/to types pair. As well as changing the logic for conversion to the suggested one

File: src/main/java/com/hazelcast/jdbc/JdbcResultSet.java (hunk 3 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:52]: Fixed

File: src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java (hunk 4 above)
Comment (pivovarit) [reply src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java:54]: It feels like we could benefit from the mutation testing integration.

Example: https://github.com/hazelcast/hazelcast-zookeeper/pull/60/files

**Review by pivovarit — COMMENTED** [review]
Looks like @devozerov covered the most important aspects already.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import java.sql.Date;
+import java.sql.SQLException;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.ZoneOffset;
+import java.util.function.Supplier;
+
+final class TypeConverter {
+
+    public static final int MILLIS_IN_SECONDS = 1_000;
```
[the remark this answers is not in the document]
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:31]: Moved it to a utility method

File: src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java (hunk 4 above)
Comment (SeriyBg) [reply src/test/java/com/hazelcast/jdbc/DriverTypeConversionTest.java:54]: Yes, definitely. I will add it as a separate PR. Also created an issue for it #19

File: src/main/java/com/hazelcast/jdbc/JdbcResultSet.java [hunk 6]
```diff
@@ -110,47 +116,47 @@ public boolean wasNull() throws SQLException {
...
     @Override
     public double getDouble(int columnIndex) throws SQLException {
-        return get(columnIndex);
+        return convertTo(get(columnIndex), QueryDataType.DOUBLE);
     }
 
     @Override
     public BigDecimal getBigDecimal(int columnIndex, int scale) throws SQLException {
-        throw JdbcUtils.unsupported("BigDecimal with scale is not supported");
+        return getBigDecimal(columnIndex).setScale(scale);
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/JdbcResultSet.java:159]: The `setScale` method may throw an `ArithmeticException`, that we do not wrap into an `SQLException`.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        SQL_TYPES_TO_QUERY_DATA_TYPE.put(Types.SMALLINT, QueryDataType.SMALLINT);
+        SQL_TYPES_TO_QUERY_DATA_TYPE.put(Types.DOUBLE, QueryDataType.DOUBLE);
+    }
+
+    private TypeConverter() {
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T convertTo(Object object, QueryDataType targetDataType) throws SQLException {
+        return convertAs(
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: All conversion methods are on the hot path. I would avoid using lambdas here because every invocation creates a new lambda instance. It is better to have some code duplication here in favor of the speed. 

Also please note that currently, we do two null checks instead of one: one check in the `convertAs` method, another in the `QueryDataType.convert`.

Another problem is that we do not consult column metadata. This forces us to call relatively heavy `Converters.getConverter` on the hot path. Instead, we may create a static table from `SqlColumnType` to `QueryDataType`, and use metadata to resolve the current column's type without `Converters.getConverter`. Also, this may lead to a situation, when, say, we have a `null ` value for the column of TYPE_A, try to convert it to TYPE_B, and it pass. While in reality, these two types are incompatible (e.g. `DATE` and `TIME` cannot be converted to each other).

To summarize, we should do a minimal number of operations on the hot path, and make sure that we throw an exception for incompatible types even if the current value is `null`.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        if (clazz == Date.class) {
+            return (T) convertToDate(object);
+        }
+        QueryDataType queryDataType = QueryDataTypeUtils.resolveTypeForClass(clazz);
+        return convertAs(object, () -> (T) queryDataType.convert(object), clazz);
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T convertTo(Object object, int targetSqlType) throws SQLException {
+        QueryDataType queryDataType = SQL_TYPES_TO_QUERY_DATA_TYPE.get(targetSqlType);
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/TypeConverter.java:75]: Why don't we support `FLOAT`, `NUMERIC`, `DECIMAL`,  `DATE`, `TIME`, `TIMESTAMP`, `NULL`, `JAVA_OBJECT`, `CHAR` types?

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    private static <T> T convertAs(Object object, Supplier<T> supplier, Class<?> targetClass) throws SQLException {
+        if (object == null) {
+            return null;
+        }
+        try {
+            return supplier.get();
+        } catch (Exception e) {
+            throw new SQLException("Cannot convert '" + object + "' of type "
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/TypeConverter.java:107]: Many tools show only the message of the top exception. In the current implementation, the user will only see that we cannot convert one type to another, but will not see the reason. We'd better add the message of the `e` to the created exception.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    static <T> T convertTo(Object object, int targetSqlType) throws SQLException {
+        QueryDataType queryDataType = SQL_TYPES_TO_QUERY_DATA_TYPE.get(targetSqlType);
+        if (queryDataType == null) {
+            throw new SQLException("Target SQL type " + targetSqlType + " is not supported");
+        }
+        return (T) queryDataType.convert(object);
+    }
+
+    static Timestamp convertToTimestamp(Object object) throws SQLException {
+        return convertAs(object, () -> new Timestamp(toMillis(
```
Comment (devozerov) [inline src/main/java/com/hazelcast/jdbc/TypeConverter.java:83]: `convertToTimestamp`, `convertToTime`, `convertToDate` - same concerns as with the `convertTo` method: allocations, heavy `Converters.getConverter`. Most of this could be avoided for a good number of cases (probably, the only exception is the source column of the `OBJECT` type).

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java (hunk 8 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:75]: AFAIK, types `DATE`, `TIME`, `TIMESTAMP`, `NULL`, `JAVA_OBJECT` are not supported at the moment by `IMDG SQL` module, please correct me if I'm wrong. So I have created the issue to add it when it will be available in `sql-module`: #26 and #25 
For all the other types will add the implementation

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java (hunk 7 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: If we are going to use `SqlColumnType` from metadata the conversion will be much more strict. For example, having an `int` field with the value `42` won't be possible to convert to `String`, for now, it is. Is this the expected behavior?

Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: Tested the behavior with `H2`, and there it is possible to call `getString` on `DECIMAL` field and conversion will be successful. And same for `null` values, having `null` as a `String` it is possible to successfully call `getBigDecimal`

Comment (devozerov) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: `H2` is a toy database, we hardly should use it as a reference. In any case, if we have a DECIMAL column, then the associated converter `BigDecimalConverter` allows for conversion to string (`BigDecimalConverter.asVarchar`).

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java (hunk 8 above)
Comment (devozerov) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:75]: They are supported in a limited way - we may get only a column of this type, but cannot use these types in expressions.

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java (hunk 7 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: Have tried with Postgres, and the behavior is the same: having a field of type `date` will `null` value I can successfully cal `getTime` which will return `null`. The same goes for `String` to `BigDecimal`. Should we follow the same behavior or do we decide to throw an exception for incompatible types even if the value can be converted?

Comment (devozerov) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: In this case, we may have a `null` check, that return `null` for every combination of types, and never fail. And only if the value is not null we resort to converters.

Comment (devozerov) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: In other words, my comment is not valid:
> Also, this may lead to a situation, when, say, we have a null value for the column of TYPE_A, try to convert it to TYPE_B, and it pass. While in reality, these two types are incompatible (e.g. DATE and TIME cannot be converted to each other).

Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:54]: Got it. So for the `null` value, we will convert anyway, otherwise will convert only if it is possible. One more interesting thing for Postgres: having a `null` value for a `String` field will return `0` for the `getLong` call. Will follow the same behavior

File: src/main/java/com/hazelcast/jdbc/TypeConverter.java (hunk 8 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/TypeConverter.java:75]: `SQL_TYPES_TO_QUERY_DATA_TYPE` is used for PreparedStatement, where it's not supported atm. For the ResultSet we do support these types

**Review by devozerov — APPROVED** [review]
